src/cosim/serialization.cpp

```diff
@@ -75,25 +86,507 @@ namespace
             out << "{\n";
             print_n_spaces(out, indent + indentStep);
             out << child->first << " = ";
-            print_tree(out, child->second, indent + indentStep, indentStep);
+            _print_ptree(out, child->second, indent + indentStep, indentStep);
             ++child;
             for (; child != tree.end(); ++child) {
                 out << ",\n";
                 print_n_spaces(out, indent + indentStep);
                 out << child->first << " = ";
-                print_tree(out, child->second, indent + indentStep, indentStep);
+                _print_ptree(out, child->second, indent + indentStep, indentStep);
             }
             out << '\n';
             print_n_spaces(out, indent);
             out << '}';
         }
     }
-}
 
+    constexpr uint64_t TYPE_INT  = 0x8000;
+    constexpr uint64_t TYPE_UINT = 0x8001;
+
+    template<typename F, typename... Args>
+    void wrap_cbor_call(F&& f, Args&&... args)
+    {
+        auto result = std::forward<F>(f)(std::forward<Args>(args)...);
+        if (!result) {
+            throw std::runtime_error("Failed Cbor operation!");
```

Comment:
Is there any way to get more info about the error and include it in the exception? It would be very useful for the user to know e.g. whether it failed because of a wrong path, insufficient permissions, etc.

If libcbor uses standard `errno` error codes, you may be able to use [`system_error`](https://en.cppreference.com/w/cpp/error/system_error/system_error) and its related functionality for this.

Reply:
Hmm unfortunately I do not see in the document that we can retrieve an error code object if the functions fail (for encoding).

Reply:
I am now passing an err message to the function so that at least we know at which stage encoding has failed.